projects/js-packages/wp-data-sync/src/create-wp-data-sync.ts

```diff
@@ -0,0 +1,361 @@
+		extractFetchResponse,
+		fetchAfterUpdate = false,
+		fetchOnInit = true,
+		getSliceFromState = defaultGetSliceFromState( name ),
+		initialState,
+		prepareUpdateRequest,
+	}: WpDataSyncOptions< Shape >
+) {
+	const capitalizedName = capitalize( name );
+
+	return {
+		resolvers: {
+			[ `get${ capitalizedName }` as const ]: () => {
+				return async function ( { dispatch } ) {
+					if ( fetchOnInit ) {
+						await dispatch[ `fetch${ capitalizedName }` ]();
+					}
+				};
+			},
+		},
+		selectors: {
+			[ `get${ capitalizedName }` as const ]: ( state: object ) => {
+				return getSliceFromState( state )?.data;
+			},
+			[ `get${ capitalizedName }Status` as const ]: ( state: object ) => {
+				return getSliceFromState( state )?.status;
+			},
+			[ `get${ capitalizedName }LastError` as const ]: ( state: object ) => {
+				return getSliceFromState( state )?.lastError;
+			},
+		},
+		actions: {
+			[ `set${ capitalizedName }` as const ]: ( payload: Partial< Shape > ) => {
+				return {
+					type: `set${ capitalizedName }` as const,
+					payload,
+				};
+			},
+			[ `setErrorFor${ capitalizedName }` as const ]: ( error: unknown ) => {
+				return {
+					type: `setErrorFor${ capitalizedName }` as const,
+					payload: {
+						error,
+					},
+				};
+			},
+			[ `setStatusFor${ capitalizedName }` as const ]: (
+				status: WpDataSyncState< Shape >[ 'status' ]
+			) => {
+				return {
+					type: `setStatusFor${ capitalizedName }` as const,
+					payload: {
+						status,
+					},
+				};
+			},
+			[ `fetch${ capitalizedName }` as const ]: () => {
+				return async function ( { dispatch, select } ) {
+					const status = select[ `get${ capitalizedName }Status` ]();
+
+					if ( status === 'fetching' || status === 'updating' ) {
+						return;
+					}
+					const setStatus = dispatch[ `setStatusFor${ capitalizedName }` ];
+
+					setStatus( 'fetching' );
+
+					try {
+						const response = await apiFetch< Response >( { path: endpoint } );
+
+						const result = extractFetchResponse?.( response ) ?? response;
+
+						dispatch[ `set${ capitalizedName }` ]( result );
+
+						setStatus( 'idle' );
+					} catch ( error ) {
+						setStatus( 'error' );
+
+						dispatch[ `setErrorFor${ capitalizedName }` ]( error );
+					}
+				};
+			},
+			[ `update${ capitalizedName }` as const ]: ( payload: Partial< Shape > ) => {
+				return async function ( { dispatch, select } ) {
+					const prevValue = select[ `get${ capitalizedName }` ]();
+
+					const setStatus = dispatch[ `setStatusFor${ capitalizedName }` ];
+
+					try {
+						if ( doOptimisticUpdates ) {
+							// Optimistically update the data.
+							dispatch[ `set${ capitalizedName }` ]( payload );
+						}
+
+						setStatus( 'updating' );
+
+						const data = prepareUpdateRequest?.( payload ) ?? payload;
+
+						await apiFetch( { method: 'POST', path: endpoint, data } );
+
+						if ( fetchAfterUpdate ) {
+							await dispatch[ `fetch${ capitalizedName }` ]();
+						}
```

Comment:
Well-written endpoints like `wp/v2/settings` return the modified settings in the POST response. You don't need to `fetchAfterUpdate`, because the data are already in the POST response. You just need to use them instead of ignoring them (i.e., not using the `await apiFetch()` return value.

This should be the default behavior. `fetchAfterUpdate` might make sense only for endpoints that are not written that well.

Also, for a really correct behavior you'll need two setter actions:
- `setSomething` is useful for merging a slice of the data into the current data
- `replaceSomething` replaces the data completely. It should be used by `fetchSomething` and also for storing the response in `updateSomething`.

If you have only `setSomething` then stale data might be stored in the state, namely the fields that were removed on the server.